plonky2x/src/frontend/eth/mpt/mpt.rs

```diff
@@ -0,0 +1,310 @@
+use std::marker::PhantomData;
+
+use curta::math::field::Field;
+use itertools::Itertools;
+
+use super::generators::*;
+use crate::prelude::{
+    ArrayVariable, BoolVariable, ByteVariable, Bytes32Variable, CircuitBuilder, PlonkParameters,
+    Variable,
+};
+
+pub fn transform_proof_to_padded<const ENCODING_LEN: usize, const PROOF_LEN: usize>(
+    storage_proof: Vec<Vec<u8>>,
+) -> (Vec<Vec<u8>>, Vec<usize>) {
+    if storage_proof.len() > PROOF_LEN {
+        panic!(
+            "Proof is too long, has {} elements, but PROOF_LEN is {}",
+            storage_proof.len(),
+            PROOF_LEN
+        );
+    }
+
+    let mut padded_elements = vec![vec![0u8; ENCODING_LEN]; PROOF_LEN];
+    let mut lengths = vec![0usize; PROOF_LEN];
+
+    for (i, inner_vec) in storage_proof.into_iter().enumerate() {
+        // Check inner length
+        if inner_vec.len() > ENCODING_LEN {
+            panic!(
+                "Proof element {} is too long, has {} elements, but ENCODING_LEN is {}",
+                i,
+                inner_vec.len(),
+                ENCODING_LEN
+            );
+        }
+        lengths[i] = inner_vec.len();
+        for (j, &byte) in inner_vec.iter().enumerate() {
+            padded_elements[i][j] = byte;
+        }
+    }
+
+    (padded_elements, lengths)
+}
+
+impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
+    pub fn le(&mut self, lhs: Variable, rhs: Variable) -> BoolVariable {
+        let generator: LeGenerator<L, D> = LeGenerator {
+            lhs: lhs.clone(),
+            rhs: rhs.clone(),
+            output: self.init::<BoolVariable>(),
+            _phantom: PhantomData,
+        };
+        self.add_simple_generator(generator.clone());
+        generator.output
+    }
+
+    pub fn byte_to_variable(&mut self, lhs: ByteVariable) -> Variable {
+        let generator: ByteToVariableGenerator<L, D> = ByteToVariableGenerator {
+            lhs: lhs.clone(),
+            output: self.init::<Variable>(),
+            _phantom: PhantomData,
+        };
+        self.add_simple_generator(generator.clone());
+        generator.output
+    }
+
+    pub fn sub_byte(&mut self, lhs: ByteVariable, rhs: ByteVariable) -> ByteVariable {
+        let generator: ByteSubGenerator<L, D> = ByteSubGenerator {
+            lhs: lhs.clone(),
+            rhs: rhs.clone(),
+            output: self.init::<ByteVariable>(),
+            _phantom: PhantomData,
+        };
+        self.add_simple_generator(generator.clone());
+        generator.output
+    }
+
+    #[allow(dead_code, unused_variables)]
+    pub fn to_nibbles(&mut self, bytes: &[ByteVariable]) -> Vec<ByteVariable> {
```

Comment:
makes more sense to impl Vec<ByteVariable>.to_nibble(&builder) probably

Reply:
will also have sidd do this